### ISSUE TITLE
Don't send 'null' when optional fields are not set

### ISSUE THREAD

**Issue by DENKweit** [infiotinc/gqlgenc#6]
Continuing the discussion from https://github.com/Yamashou/gqlgenc/issues/64

Given this schema:
```
input GetFooInput {
    optionalParam1: String
    optionalParam2: String
}

type Query {
    getFoo(input: GetFooInput!): String!
}
```

this operation:
```
query getFoo($p1: String, $p2: String) {
  getFoo(input: {optionalParam1: $p1, optionalParam2: $p2})
}
```

and this example code:
```golang
package main

import (
	"context"
	"lol/dgraphclient"

	"github.com/infiotinc/gqlgenc/client"
	"github.com/infiotinc/gqlgenc/client/transport"
)

func main() {
	wstr := &transport.Ws{
		URL: "ws://localhost:8686/subscribe",
	}
	wstr.Start(context.Background())
	defer wstr.Close()

	httptr := &transport.Http{
		URL: "http://localhost:8686/subscribe",
	}

	tr := transport.SplitSubscription(wstr, httptr)

	cli := &client.Client{
		Transport: tr,
	}

	c := dgraphclient.Client{Client: cli}

	p1 := "hi"

	c.GetFoo(context.Background(), &p1, nil)
}
```

Following request is being made to the server:
```json
{
  "query": "query getFoo ($p1: String, $p2: String) {\n\tgetFoo(input: {optionalParam1:$p1,optionalParam2:$p2})\n}\n",
  "operationName": "GetFoo",
  "variables": { "p1": "hi", "p2": null }
}

```

This is problematic, because setting `p2` to `null` is different than not sending it in the first place.
Sending `null` can (and will) invoke actions on the server side. E.g. deleting some database entry.

Therefore it would be great if we could have the concept of `undefined` in this golang client, which is of course difficult as golang does not support an `undefined` type.

# Possible Solution

Maybe something like this could work:

```golang
type OptionalString struct {
	Value *string
}

type GetFooInput struct {
        OptionalParam1 *OptionalString `json:"optionalParam1,omitempty"`
	OptionalParam2 *OptionalString `json:"optionalParam2,omitempty"`
}
```
I.e. when optional input types are present (on any depth) an `Optional<Type>` type is used in combination with jsons `omitempty` tag instead. This behavior could be made configurable via `.gqlgenc.yml`.

**Case A**:
When the field is `nil`, it will not be serialized.
**Case B**:
When it is set to `&OptionalString{}`, the client sends `null`.
**Case C**:
When it is set to `&OptionalString{Value: "foo"}`, the client sends `"foo"`.

Here is an working example:
```golang
package main

import (
	"encoding/json"
	"fmt"
)

type OptionalString struct {
	Value *string
}

func (o *OptionalString) MarshalJSON() ([]byte, error) {
	return json.Marshal(o.Value)
}

type GetFooInput struct {
	OptionalParam1 *OptionalString `json:"optionalParam1,omitempty"`
	OptionalParam2 *OptionalString `json:"optionalParam2,omitempty"`
}

func main() {
	s := "foo"
	f := GetFooInput{
		OptionalParam1: &OptionalString{Value: &s},
	}

	result, err := json.Marshal(f)
	if err != nil {
		panic(err)
	}

	fmt.Println(string(result))
}

```

Which outputs:
```
{"optionalParam1":"foo"}
```

It would be great if you could add this feature!

**Comment by DENKweit** [comment]
I implemented this [here](https://github.com/DENKweit/gqlgenc).

If you like, have a look. I can also create a PR and we can collaborate on final touches.

**Comment by raphaelfff** [comment]
Thank you for your report!
I was thinking about a more scalable solution:

I think we should generate the required params in the signature, and let the optional ones as [Option pattern](https://www.sohamkamani.com/golang/options-pattern/), something like that:

```go
interface GetFooParam {
    setGetFooParam(k string, v interface{})
}

type getFooParam struct {
    key string
    value interface{}
}

func OptionalParam1(v string) GetFooParam{
    return getFooParam{
        key: "optionalParam1",
        value: v,
    }
}

func (c Client) GetFoo(ctx context.Context, requiredParam int, params...GetFooParam) {
...
}
```

That would be used this way:

```go

_, _ = c.GetFoo(ctx, 10, OptionalParam1("some value"))

```

**Comment by raphaelfff** [comment]
Let me know what you think about that solution

**Comment by DENKweit** [comment]
Hi, 

input parameters can be optional on arbitrary depths:

```
input FooInput {
   bar: BarInput
}

input BarInput {
  baz: BazInput {
}

input BazInput {
  name: String
}

type Query {
  test($input: FooInput!): String
}
```

In fact, for a quick test, I already replaced all occurences of Yamashou/gqlgenc in our backend (10+ services) with our fork and it didn't feel awkward to use IMHO.

Could you elaborate on why this solution would not be "scalable"?

**Comment by raphaelfff** [comment]
So maybe I didnt understand your changes correctly, but it seems that you have hardcoded the existence of `Optional` types for `int`, `string` etc [here](https://github.com/DENKweit/gqlgenc/commit/a6fa296ec1f0b902695139f919ed2ebc1b7e2fea#diff-0e6b55bc2f2679c6d40476cbf6429a3c75dc5dd9ecbae9b4a38ef47aba5e61a2R168)

What if i want an OptionalFooInput ?

Also, thinking about it a bit more, in GQL, or something exist (non null), or it doesnt (null), so the issue IMO is that the vars should only be marshaled to JSON if they are not null ([those vars] for example)(https://github.com/infiotinc/gqlgenc/blob/master/example/client/gen_client.go#L177). In other words, the vars and input type should have the `omitempty` tag if they are nullable.

**Comment by DENKweit** [comment]
@raphaelfff  that was my first attempt but in the most recent version its all auto-generated. Nothing hardcoded anymore. And `OptionalFooInput` is therefore also absolutely possible.

**Comment by raphaelfff** [comment]
My solution has one case not covered tho: we need a specific field to be sent as null, as opposed to not being sent...
Ah, could you point me to the fully generated version of it ?

**Comment by DENKweit** [comment]
The important commits can be found [here](https://github.com/DENKweit/gqlgenc/commit/cf925efbe162838f8279f469aa40b86861e800b7) and [here](https://github.com/DENKweit/gqlgenc/commit/ff77787cc1261728368db446953275d12688d2c6).

The latest master has everything working.

**Comment by raphaelfff** [comment]
Thank you!
A few points here:
- As you can see in #3 , I'm trying to move away as much as possible from the gqlgen/modelgen
- GQL should be encoding agnostic, (not tied to json) so i'm trying to limit the amount of references to json, so that its easy to implements a custom marshal if its ever needed (some crazy people might want to have it use messagepack ^^)

The use of `Optional` types makes it a bit more cumberstone to use, where its not necessary, although some places may require it (ilke in your use case), so I was thinking about triggering the use of `Optional` through a directive, what do you think about it ?

**Comment by DENKweit** [comment]
Yes, I agree. Configurable per config flag would be totally fine.

It could be enabled per default though as I don't see why one would not want to use it.
Besides the fact that sending `null` might trigger some action on server-side you also save loooots of bandwidth and your server spends less time on decoding json/(anything else).

What is the reason for moving away from gqlgen/modelgen? Only the JSON part?
I agree on your second point. One could build a plugin system to allow arbitrary encoding libs. JSON could be default though.

**Comment by raphaelfff** [comment]
Agreed on the space saved & decoding time, but point against it enabled by default is that it forces you to wrap all your `strings` with some `Optional`,  turning a `getFoo(&"yolo")` into `getFoo(&OptionalString("yolo"))`. (yeah you cant directly take a pointer from a string, but you got the gist).
I m not heavily against it, i just think it needs some reflection

**Comment by raphaelfff** [comment]
Coming back to the Options pattern, its more elegant to use,  but requires a LOT of codegen if we support nested optional fields this way...

**Comment by DENKweit** [comment]
>Coming back to the Options pattern, its more elegant to use, but requires a LOT of codegen if we support nested optional fields this way...
Which conflicts with the "elegance", right?

I understanding the options pattern like this:
```
```
```golang
FooParams().With
```

**Comment by raphaelfff** [comment]
That is an interesting question... it is a bit more elegant, but indeed makes auto complete less useful. Unfortunately with go thare are no perfect solutions...
For the `null` case... i have a dirty idea, for nullable fields, generate it this way:
```go
func (i fooInput) WithBar(bar ...barInput) fooInput {
        if len(bar) > 0 {
                 i.params["bar"] = bar[0]
        } else {
                 i.params["bar"] = null
        }
	
	return i
}
```

I guess that options pattern would work, and emulates the full capabilities of JSON while keeping it somewhat easy to use. I think that behavior should be enabled in the config (per field or globally to account for people's needs; for ex, in my case, we never had a need for that specific nullity, so using direct objects is way easier).

**Comment by raphaelfff** [comment]
> What is the reason for moving away from gqlgen/modelgen? Only the JSON part?

I realise i forgot to answer that; gqlgen/modelgen generates the types for all existing GQL types, regardless of whether they are  used or not; which defeats one of the points of using the queries to generate code, that's what #3 is trying to address

**Comment by DENKweit** [comment]
>For the null case... i have a dirty idea, for nullable fields, generate it this way:

I also thought about this but this would allow to supply multiple params and unless we can somehow statically assert that `len(bar) <= 1`. Which we can't afaik. So this would be bad practice IMHO.

>EDIT: I'm not sure what's best between the options pattern and the Optional struct...
I think I'm fine with both although using the `optional`-struct variant would require less code changes in case we migrate.

>I realise i forgot to answer that; gqlgen/modelgen generates the types for all existing GQL types, regardless of whether they are used or not; which defeats one of the points of using the queries to generate code, that's what #3 is trying to address.

Which is a problem for auto-completion only, right? At compile time unused structs would be removed anyway.

>Also, gqlgen/modelgen is pretty tricky to hook into, to implement the Optional structs, you have to copy/pasta their code and modify it :/ , also I feel like we dont really need that modelgen, since its targeted towards server, most of the gen for gqlgenc is done in gqlgenc, I feel like the same should be for input types.

Yes, agreed.

**Comment by raphaelfff** [comment]
> I also thought about this but this would allow to supply multiple params and unless we can somehow statically assert that len(bar) <= 1 - which we can't afaik - this is bad practice IMHO.
Indeed, an interesting hack, not a reliable solution imo

> Which is a problem for auto-completion only, right? At compile time unused structs would be removed anyway.
Yeah, it also is unnecessary, and depending on the size of the API, can lead to generating a LOT of code being unused, also just adds noise to PRs

**Comment by DENKweit** [comment]
Sooo... How to proceed? :laughing:

**Comment by raphaelfff** [comment]
I m currently evaluating in #3 what to do with the modelgen, i think i will partially integrate it in the codebase of gqlgenc, and split it out into util functions. Not a huge fan of copy/pasta, but its the only solution.

Your suggestion seems good, we actually need to generate `New<type>Ptr`, to accomodate for scalars and custom types(for ex `type MySuperString string`).
On top of that, we also need to generate a custom JSONMarshal for each input type

**Comment by DENKweit** [comment]
>Your suggestion seems good, we actually need to generate New<type>Ptr, to accomodate for scalars and custom types(for ex type MySuperString string).
Yes, doesn't hurt to generate for basic types also. Agreed.

>On top of that, we also need to generate a custom JSONMarshal for each input type
Do we? I think built-in serialization of `map[string]interface{}` might be working just fine already.

By the way, I found what I didn't like about the pointer solution. But it's only a minor:
```golang
// generated func for required input argument
func (c*client) GetFoo(ctx context.Context, input fooInput) { ... }

// usage
_, _ = c.GetFoo(ctx, *NewFooInput(...)) // note the *
```

I think that this is OK though. And it would also allow easily for optional arguments on the root level:
```golang
func (c*client) GetFoo(ctx context.Context, input * fooInput) { ... }
_, _ = c.GetFoo(ctx, nil) 
```

**Comment by raphaelfff** [comment]
It wont because params is a private field, also its one level too deep.

Instead, I think we should do something like that:

```go
type fooInput map[string]interface{}

func NewFooInput(required string) *fooInput {
   return fooInput(map[string]interface{}{
     "required": required,
    })
}

func (i *fooInput) WithBar(bar *barInput) *fooInput {
	i["bar"] = bar
	return i
}

func (i *fooInput) WithValue(value *int) *fooInput {
	i["value"] = value
	return i
}

```

That would also mean that this feature would be called "input as map" which is the recommended way in go to handle such case (with json and nullity); we just generate some syntactic sugar/helpers

**Comment by DENKweit** [comment]
Perfect!

**Comment by raphaelfff** [comment]
@DENKweit Would you be able to give a try to whatever is in #10 ? Works for us, I wanna make sure I didn't forget anything before merging

**Comment by DENKweit** [comment]
Yes, will try tomorrow and let you know!

**Comment by raphaelfff** [comment]
@DENKweit Did you get the opportunity to give it a go ?

**Comment by DENKweit** [comment]
I just tried it out but I'm a bit confused. Did you implement any of the optional pattern discussed above?

**Comment by raphaelfff** [comment]
#10 does not have the input as map, this one is just about removing modelgen, the input as map is basically ready, not pushed yet

**Comment by DENKweit** [comment]
Oh, I'm totally fine with removing modelgen. Go ahead! :)

**Comment by raphaelfff** [comment]
@DENKweit If you want to give it a spin, input as map is available in #11

**Comment by raphaelfff** [comment]
Closed in #11
